File: backend/featurePacks/huggingfacePack.ts
export interface HuggingFaceFile {
  path: string;
  size: number;
  type: 'file' | 'directory';
  downloadUrl: string;
  isWeight: boolean; // .safetensors, .bin, .onnx, .pt, .gguf
}

export interface HuggingFaceModelInfo {
  modelId: string;
  author: string;
  repoName: string;
  revision: string;
  pipelineTag?: string;
  files: HuggingFaceFile[];
  totalWeightSize: number;
}

export class HuggingFacePack {
  private userAgent = 'VortexDownloader-HuggingFacePack/1.0.0';

  /**
   * Parse a Hugging Face URL
   */
  public parseUrl(urlStr: string): {
    isHuggingFace: boolean;
    modelId?: string;
    author?: string;
    repoName?: string;
    revision?: string;
    filePath?: string;
    directUrl?: string;
    isDirectFile?: boolean;
  } {
    try {
      const parsed = new URL(urlStr);
      const host = parsed.hostname.toLowerCase();

      if (host !== 'huggingface.co' && host !== 'hf.co') {
        return { isHuggingFace: false };
      }

      const parts = parsed.pathname.split('/').filter(Boolean);
      if (parts.length < 1) {
        return { isHuggingFace: false };
      }

      // Special prefixes: datasets/..., spaces/...
      const isDataset = parts[0] === 'datasets';
      const offset = isDataset ? 1 : 0;

      if (parts.length < offset + 1) {
        return { isHuggingFace: false };
      }

      let author = '';
      let repoName = '';
      let restIndex = 0;

      if (parts.length === offset + 1) {
        // Single name model, e.g., /gpt2 or /bert-base-uncased
        author = 'openai-community'; // or generic
        repoName = parts[offset];
        restIndex = offset + 1;
      } else {
        author = parts[offset];
        repoName = parts[offset + 1];
        restIndex = offset + 2;
      }

      const modelId = `${author}/${repoName}`;

      // Direct file link: /resolve/:revision/... or /raw/:revision/...
      if (
        parts.length > restIndex + 2 &&
        (parts[restIndex] === 'resolve' || parts[restIndex] === 'raw')
      ) {
        const revision = parts[restIndex + 1];
        const filePath = parts.slice(restIndex + 2).join('/');
        const directUrl = `https://huggingface.co/${modelId}/resolve/${revision}/${filePath}`;
        return {
          isHuggingFace: true,
          modelId,
          author,
          repoName,
          revision,
          filePath,
          directUrl,
          isDirectFile: true
        };
      }

      // Tree view: /tree/:revision
      let revision = 'main';
      if (parts.length > restIndex + 1 && parts[restIndex] === 'tree') {
        revision = parts[restIndex + 1];
      }

      return {
        isHuggingFace: true,
        modelId,
        author,
        repoName,
        revision,
        isDirectFile: false
      };
    } catch (_) {
      return { isHuggingFace: false };
    }
  }

  /**
   * Inspect a model or dataset repo and return all files
   */
  public async inspectModel(modelId: string, revision = 'main'): Promise<HuggingFaceModelInfo> {
    const parts = modelId.split('/');
    const author = parts.length > 1 ? parts[0] : '';
    const repoName = parts.length > 1 ? parts[1] : modelId;

    // Use HF API tree endpoint
    const endpoint = `https://huggingface.co/api/models/${modelId}/tree/${encodeURIComponent(revision)}`;

    const res = await fetch(endpoint, {
      headers: {
        'User-Agent': this.userAgent,
        'Accept': 'application/json'
      }
    });

    if (!res.ok) {
      throw new Error(`Failed to inspect Hugging Face repo (${res.status} ${res.statusText})`);
    }

    const items: any[] = await res.json();
    const weightExtensions = ['.safetensors', '.bin', '.onnx', '.pt', '.pth', '.gguf', '.h5', '.msgpack'];

    let totalWeightSize = 0;
    const files: HuggingFaceFile[] = items
      .filter((item) => item.type === 'file')
      .map((item) => {
        const lower = item.path.toLowerCase();
        const isWeight = weightExtensions.some((ext) => lower.endsWith(ext));
        const size = typeof item.size === 'number' ? item.size : 0;
        if (isWeight) totalWeightSize += size;

        return {
          path: item.path,
          size,
          type: 'file',
          downloadUrl: `https://huggingface.co/${modelId}/resolve/${revision}/${item.path}`,
          isWeight
        };
      });

    // Sort so weights and major model files are at the top
    files.sort((a, b) => {
      if (a.isWeight && !b.isWeight) return -1;
      if (!a.isWeight && b.isWeight) return 1;
      return b.size - a.size;
    });

    return {
      modelId,
      author,
      repoName,
      revision,
      files,
      totalWeightSize
    };
  }
}

File: backend/feature_packs/huggingface_pack.py
import re
import httpx
from urllib.parse import urlparse
from typing import Dict, Any, Optional

class HuggingFacePack:
    def parse_url(self, raw_url: str) -> Dict[str, Any]:
        parsed = urlparse(raw_url)
        hostname = (parsed.hostname or '').lower()

        if hostname != 'huggingface.co' and hostname != 'hf.co':
            return {'isHuggingFace': False}

        path = parsed.path.strip('/')
        parts = path.split('/')

        if len(parts) < 2:
            return {'isHuggingFace': False}

        # Resolve direct resolve URL
        # e.g. /TheBloke/Llama-2-7B-GGUF/resolve/main/llama-2-7b.Q4_K_M.gguf
        if 'resolve' in parts:
            idx = parts.index('resolve')
            model_id = "/".join(parts[:idx])
            revision = parts[idx + 1] if idx + 1 < len(parts) else 'main'
            file_path = "/".join(parts[idx + 2:])
            return {
                'isHuggingFace': True,
                'isDirectFile': True,
                'modelId': model_id,
                'revision': revision,
                'filePath': file_path,
                'directUrl': raw_url
            }

        # Repository URL
        model_id = f"{parts[0]}/{parts[1]}"
        return {
            'isHuggingFace': True,
            'isDirectFile': False,
            'modelId': model_id,
            'revision': 'main'
        }

    async def inspect_model(self, model_id: str, revision: str = 'main') -> Dict[str, Any]:
        api_url = f"https://huggingface.co/api/models/{model_id}"
        headers = {'User-Agent': 'VortexDownloader/0.6.1'}

        async with httpx.AsyncClient(timeout=10.0) as client:
            res = await client.get(api_url, headers=headers)
            if res.status_code != 200:
                raise Exception(f"Hugging Face API error: HTTP {res.status_code}")
            data = res.json()

            siblings = []
            for s in data.get('siblings', []):
                fn = s.get('rfilename')
                if fn:
                    siblings.append({
                        'rfilename': fn,
                        'downloadUrl': f"https://huggingface.co/{model_id}/resolve/{revision}/{fn}"
                    })

            return {
                'id': data.get('id'),
                'author': data.get('author'),
                'downloads': data.get('downloads', 0),
                'likes': data.get('likes', 0),
                'pipeline_tag': data.get('pipeline_tag'),
                'files': siblings
            }
